### ISSUE TITLE
[WinError 10053] with the 0.21.0

### ISSUE THREAD

**Issue by Oxyo78** [mozilla/geckodriver#1305]
## System

* Version: 0.21.0 (64 bits)
* Platform: Windows10 but using bash(Ubuntu)
* Firefox: 60.0.2 (64 bits)
* Selenium: 3.12.0

## Testcase

```
from selenium.webdriver.firefox.webdriver import WebDriver
from selenium.webdriver.common.keys import Keys

class MySeleniumTests(StaticLiveServerTestCase):
    """ Try connexion and logout of a user """
    fixtures = ['user.json', 'categorie.json', 'product.json']
    from selenium.webdriver.support.wait import WebDriverWait
    timeout = 2

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls.selenium = WebDriver()
        cls.selenium.implicitly_wait(10)

    @classmethod
    def tearDownClass(cls):
        cls.selenium.quit()
        super().tearDownClass()

    def test_login(self):
        """ Login & disconnect user """
        from selenium.webdriver.support.wait import WebDriverWait
        timeout = 2
        self.selenium.get('%s%s' % (self.live_server_url, '/connexion/'))
        username_input = self.selenium.find_element_by_name("username")
        username_input.send_keys('Paul')
        password_input = self.selenium.find_element_by_name("password")
        password_input.send_keys('azerty')
        self.selenium.find_element_by_xpath(
            '//input[@value="Se connecter"]').click()
        WebDriverWait(self.selenium, timeout).until(
            lambda driver: driver.find_element_by_id('carrot'))
        self.selenium.find_element_by_xpath(
            '//a[@title="Déconnexion"]').click()

```

## Output

`ConnectionAbortedError: [WinError 10053] An established connection was aborted by the software in your host machine`

Hi, this is my selenium test under Django with the last version of geckodriver (0.21.0 64bits), I got this error when I run the test. But when I use the geckodriver version (0.20.1 64bits), It's working fine.
Anyone can tell me what's wrong with my test and last geckodriver version?

**Comment by SkNuwanTissera** [comment]
Sent from Mail for Windows 10

From: Yohan V
Sent: Sunday, June 24, 2018 5:12 AM
To: mozilla/geckodriver
Cc: Subscribed
Subject: [mozilla/geckodriver] [WinError 10053] with the 0.21.0 (#1305)

System
• Version: 0.21.0 (64 bits)
• Platform: Windows10 but using bash(Ubuntu)
• Firefox: 60.0.2 (64 bits)
• Selenium: 3.12.0
Testcase
from selenium.webdriver.firefox.webdriver import WebDriver
from selenium.webdriver.common.keys import Keys
class MySeleniumTests(StaticLiveServerTestCase):
""" Try connexion and logout of a user """
fixtures = ['user.json', 'categorie.json', 'product.json']
from selenium.webdriver.support.wait import WebDriverWait
timeout = 2
@classmethod
def setUpClass(cls):
    super().setUpClass()
    cls.selenium = WebDriver()
    cls.selenium.implicitly_wait(10)

@classmethod
def tearDownClass(cls):
    cls.selenium.quit()
    super().tearDownClass()

def test_login(self):
    """ Login & disconnect user """
    from selenium.webdriver.support.wait import WebDriverWait
    timeout = 2
    self.selenium.get('%s%s' % (self.live_server_url, '/connexion/'))
    username_input = self.selenium.find_element_by_name("username")
    username_input.send_keys('Paul')
    password_input = self.selenium.find_element_by_name("password")
    password_input.send_keys('azerty')
    self.selenium.find_element_by_xpath(
        '//input[@value="Se connecter"]').click()
    WebDriverWait(self.selenium, timeout).until(
        lambda driver: driver.find_element_by_id('carrot'))
    self.selenium.find_element_by_xpath(
        '//a[@title="Déconnexion"]').click()
Output
ConnectionAbortedError: [WinError 10053] An established connection was aborted by the software in your host machine
Hi, this is my selenium test under Django with the last version of geckodriver (0.21.0 64bits), I got this error when I run the test. But when I use the geckodriver version (0.20.1 64bits), It's working fine.
Anyone can tell me what's wrong with my test and last geckodriver version?

**Comment by whimboo** [comment]
Can you both please provide a trace level log? See https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/TraceLogs.html in how to get one.

I assume this issue is a dupe of #1304 but I need more details first.

**Comment by Oxyo78** [comment]
This is the fresh trace log
[geckodriver.log](https://github.com/mozilla/geckodriver/files/2136456/geckodriver.log)

**Comment by whimboo** [comment]
@Oxyo78 nothing stands out there. Mind also attaching the Selenium log?

Also does any of the commands you execute here take longer than 5s to execute? Or maybe caused by an extra sleep?

**Comment by Oxyo78** [comment]
This the log of one my selenium test:
```
ERROR: test_login (food.tests.MySeleniumTests)
Login & disconnect user
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\Yohan\Documents\Python_project\PureButter11\food\tests.py", line 101, in test_login
    self.selenium.get('%s%s' % (self.live_server_url, '/connexion/'))
  File "C:\Users\Yohan\Documents\Python_project\PureButter11\env\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 326, in get
    self.execute(Command.GET, {'url': url})
  File "C:\Users\Yohan\Documents\Python_project\PureButter11\env\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 312, in execute
    response = self.command_executor.execute(driver_command, params)
  File "C:\Users\Yohan\Documents\Python_project\PureButter11\env\lib\site-packages\selenium\webdriver\remote\remote_connection.py", line 472, in execute
    return self._request(command_info[0], url, body=data)
  File "C:\Users\Yohan\Documents\Python_project\PureButter11\env\lib\site-packages\selenium\webdriver\remote\remote_connection.py", line 495, in _request
    self._conn.request(method, parsed_url.path, body, headers)
  File "C:\Users\Yohan\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "C:\Users\Yohan\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "C:\Users\Yohan\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "C:\Users\Yohan\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 1065, in _send_output
    self.send(chunk)
  File "C:\Users\Yohan\AppData\Local\Programs\Python\Python36-32\lib\http\client.py", line 986, in send
    self.sock.sendall(data)
ConnectionAbortedError: [WinError 10053] Une connexion établie a été abandonnée par un logiciel de votre ordinateur hôte
```

**Comment by whimboo** [comment]
Is there any change to see some timing information in the Selenium log? The information above is not that verbose.

**Comment by Oxyo78** [comment]
I tried to get the selenium log, but doesn't work.

**Comment by westdark** [comment]
As suggested by https://stackoverflow.com/questions/50996452/python-selenium-winerror-10053-host-is-shutting-connection-down, I ran into this problem with geckodriver 0.21.0, but not 0.20.1.

**Comment by Oxyo78** [comment]
Thx for reply, I know this post, it was my answer :)
I don't know why Selenium doesn't want to give me his logs. Maybe something wrong in my code, but still working with the 0.20.1.
Thx to taking time to check my problem

**Comment by whimboo** [comment]
@Oxyo78 please also answer my other question as asked in https://github.com/mozilla/geckodriver/issues/1305#issuecomment-400240581. Thanks.

**Comment by Oxyo78** [comment]
@whimboo I'm sorry, I forgot to answer. the both version take same time, but the 0.21.0 give me the error.
I found  where the error come from.

Like this, I got the error:
`class MySeleniumTests(StaticLiveServerTestCase):
    """ Try connexion and logout of a user """
    fixtures = ['user.json', 'categorie.json', 'product.json']

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        opts = Options()
        opts.log.level = "trace"
        opts.set_headless(headless=True)
        cls.selenium = webdriver.Firefox()
        cls.selenium.implicitly_wait(10)

    @classmethod
    def tearDownClass(cls):
        cls.selenium.quit()
        super().tearDownClass()

    def test_local(self):
        self.selenium.get("http://www.python.org")
        assert "Python" in self.selenium.title`

the function "test_local" doesn't want to get the `cls.selenium = webdriver.Firefox()`
But if I set it in the function like this, it's working:

`def test_local(self):
    self.selenium = webdriver.Firefox()
    self.selenium.get("http://www.python.org")
    assert "Python" in self.selenium.title`

**Comment by Oxyo78** [comment]
You can find my project here, if you want to try:
https://github.com/Oxyo78/PureButter11

**Comment by Oxyo78** [comment]
Finally I fix it with the 0.21.0.
I changed my setUp and tearDown function name and code like this:
```
    def setUp(self):
        opts = Options()
        opts.log.level = "trace"
        opts.set_headless(headless=True)
        self.selenium = webdriver.Firefox(firefox_options=opts)

    def tearDown(self):
        self.selenium.quit()

    def test_local(self):
        self.selenium.get("http://www.python.org")
        assert "Python" in self.selenium.title
```
I've still get the [WinError 10053] error, but all the test is OK

**Comment by whimboo** [comment]
Hm, mind uploading another trace log which shows such an error? I would still be interested to know what this actually is. Thanks.

**Comment by Oxyo78** [comment]
[geckodriver.log](https://github.com/mozilla/geckodriver/files/2150708/geckodriver.log)
This is the new trace log with the version 0.21.0

**Comment by whimboo** [comment]
@Oxyo78 please note that the log will always be appended. So what you lately attached is huge. It's not clear to me where it failed. Mind deleting the file and running it again? Maybe even with a debug build of Firefox Nightly? You can grab the binary here: https://queue.taskcluster.net/v1/task/Wi2cj0usRIuh53v0B7dnOQ/runs/0/artifacts/public/build/target.zip. Thanks.

**Comment by whimboo** [comment]
@Oxyo78 I actually don't have the time to try running your project locally, but I have an idea. I assume that this behavior could be related to the enabling of HTTP keep-alive connections. There is also issue #1304 for that but with a different failure message. Maybe it differs between Linux and Windows systems.

What makes me wonder now are you using keep-alive? By default it should be disabled for the Python bindings as @AutomatedTester told me. So if you explicitly enable it, try disabling it. 

Otherwise please reduce the implicit timeout from 10s to something <5s. Does that also help? If yes, it should be the keep-alive problem, given that at the moment the connection gets closed after 5s of inactivity.

**Comment by AutomatedTester** [comment]
@whimboo This issue is because the connection is being forcibly closed by a webserver.

From a cursory search there is an issue with django or flask apps not being able to handle concurrent connections when not being put through a proper WSGI server. See https://stackoverflow.com/questions/12591760/flask-broken-pipe-with-requests

**Comment by gtkpr** [comment]
i'm having the exact same issue on Windows 10 all of a sudden with geckdriver version 0.21.0.

```
Traceback (most recent call last):
  File "c:\python27\lib\site-packages\twisted\internet\defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "C:\Users\Anthony\Desktop\scrapy\projects\crowdfunding\spiders\indiegogo.py", line 537, in parse
    self.scrape_details_of_each_campaign(funding)
  File "C:\Users\Anthony\Desktop\scrapy\projects\crowdfunding\spiders\indiegogo.py", line 274, in scrape_details_of_each_campaign
    backers_profile = container.find_elements_by_xpath('./*/div[@class="campaignBackers-pledge-backer-details"]/a')
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\webelement.py", line 381, in find_elements_by_xpath
    return self.find_elements(by=By.XPATH, value=xpath)
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\webelement.py", line 678, in find_elements
    {"using": by, "value": value})['value']
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\webelement.py", line 628, in _execute
    return self._parent.execute(command, params)
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\webdriver.py", line 312, in execute
    response = self.command_executor.execute(driver_command, params)
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\remote_connection.py", line 472, in execute
    return self._request(command_info[0], url, body=data)
  File "c:\python27\lib\site-packages\selenium\webdriver\remote\remote_connection.py", line 496, in _request
    resp = self._conn.getresponse()
  File "c:\python27\lib\httplib.py", line 1121, in getresponse
    response.begin()
  File "c:\python27\lib\httplib.py", line 438, in begin
    version, status, reason = self._read_status()
  File "c:\python27\lib\httplib.py", line 394, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "c:\python27\lib\socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
error: [Errno 10053] An established connection was aborted by the software in your host machine
```

**Comment by whimboo** [comment]
@Oxyo78, Selenium 3.14.0 just got released. Can you please upgrade and check again if it is fixed for you? Thanks

**Comment by FelipeGdM** [comment]
I had the same error with Selenium 3.13 and the new version 3.14 fixed it

**Comment by gtkpr** [comment]
@FelipeGdM i hope that fixes it!

**Comment by whimboo** [comment]
Great to hear. So marking this issue as dupe of #1304.

**Comment by Oxyo78** [comment]
Hello, I'm still busy at this moment, I will try thisnew version on this weekend and give back

**Comment by whimboo** [comment]
As it looks like this is still a present issue. As such I will reopen this issue.

**Comment by whimboo** [comment]
Everyone who is experiencing this issue, could you please try the following nightly build of geckodriver? It has some improvements compared to the 0.21.0 release, which might help in this particular case. You can download it from https://queue.taskcluster.net/v1/task/BC3gJ9L8Sne6Psppd5cVmw/runs/0/artifacts/public/build/target.common.tests.tar.gz. After unpacking please find the binary under `bin/geckodriver.exe`.

Please let me know how this works for you. Thanks.

**Comment by GraceAredel** [comment]
@whimboo Hello Henrik! Thank you very much for the new build you provided! It fixed my issue.

**Comment by whimboo** [comment]
Ok, so that's good to know. It means that it got fixed by https://bugzilla.mozilla.org/show_bug.cgi?id=1471201. But note that every sleep or duration higher than 60s might still cause this problem.

@GraceAredel could you test with an asynchronous script and a timer which resolves the script after 60s? Thanks.

**Comment by whimboo** [comment]
Looks like I missed to close this issue, which got fixed in the geckodriver 0.22.0 release.

**Comment by GraceAredel** [comment]
@whimboo Oh! I'm so sorry that I forgot to answer you! I wrote the script with 60s wait before resolving and it worked. Sorry again!
